lib/receive.dart

```diff
@@ -202,78 +158,157 @@ class MaybeInputsSeen extends FfiV2MaybeInputsSeen {
+
+  @override
+  Future<bool> isOutputSubstitutionDisabled({hint}) {
+    try {
+      return super.isOutputSubstitutionDisabled();
+    } on error.PayjoinError catch (e) {
+      throw mapPayjoinError(e);
+    }
+  }
 
   @override
-  Future<void> trySubstituteReceiverOutput(
-      {required FutureOr<Uint8List> Function() generateScript}) {
+  Future<WantsOutputs> replaceReceiverOutputs(
+      {required List<TxOut> replacementOutputs,
+      required FfiScript drainScript}) async {
     try {
-      return super.trySubstituteReceiverOutput(generateScript: generateScript);
+      final res = await super.replaceReceiverOutputs(
+          replacementOutputs: replacementOutputs, drainScript: drainScript);
+      return WantsOutputs._(field0: res.field0);
     } on error.PayjoinError catch (e) {
       throw mapPayjoinError(e);
     }
   }
 
   @override
-  Future<void> contributeWitnessInput(
-      {required TxOut txo, required OutPoint outpoint, hint}) {
+  Future<WantsOutputs> substituteReceiverScript(
+      {required FfiScript outputScript}) async {
     try {
-      return super.contributeWitnessInput(txo: txo, outpoint: outpoint);
+      final res =
+          await super.substituteReceiverScript(outputScript: outputScript);
+      return WantsOutputs._(field0: res.field0);
     } on error.PayjoinError catch (e) {
       throw mapPayjoinError(e);
     }
   }
 
-  /// Select receiver input such that the common.dart avoids surveillance. Return the input chosen that has been applied to the Proposal.
+  @override
+  Future<WantsInputs> commitOutputs() async {
+    try {
+      final res = await super.commitOutputs();
+      return WantsInputs._(field0: res.field0);
+    } on error.PayjoinError catch (e) {
+      throw mapPayjoinError(e);
+    }
+  }
+}
+
+class WantsInputs extends FfiWantsInputs {
+  WantsInputs._({required super.field0});
+
+  /// Select receiver input such that the payjoin avoids surveillance.
+  /// Return the input chosen that has been applied to the Proposal.
   ///
-  /// Proper coin selection allows common.dart to resemble ordinary transactions. To ensure the resemblance, a number of heuristics must be avoided.
+  /// Proper coin selection allows payjoin to resemble ordinary transactions.
+  /// To ensure the resemblance, a number of heuristics must be avoided.
   ///
-  /// UIH “Unnecessary input heuristic” is one class of them to avoid. We define UIH1 and UIH2 according to the BlockSci practice BlockSci UIH1 and UIH2:
+  /// UIH "Unnecessary input heuristic" is one class of them to avoid. We define
+  /// UIH1 and UIH2 according to the BlockSci practice
+  /// BlockSci UIH1 and UIH2:
+  /// if min(out) < min(in) then UIH1 else UIH2
+  /// https://eprint.iacr.org/2022/589.pdf
   @override
-  Future<OutPoint> tryPreservingPrivacy(
-      {required Map<BigInt, OutPoint> candidateInputs, hint}) {
+  Future<InputPair> tryPreservingPrivacy(
+      {required List<FfiInputPair> candidateInputs}) async {
```

Comment:
Better to use InputPair instead of FfiInputPair to not have internal types in the external functions.

Reply:
because the internal super.tryPreservingPrivacy takes a List<FfiInputPair>, using List<InputPair> isn't a valid override